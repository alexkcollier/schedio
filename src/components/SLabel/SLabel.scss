.label {
  @include font-size(5);
  display: block;
  font-weight: $font-weight-bold;
  text-transform: uppercase;

  &--inline {
    align-items: center;
    display: inline-flex;
    flex-direction: row;
    justify-content: flex-start;

    .input,
    .textarea,
    .select {
      margin-left: spacing();
    }

    &.label--reversed {
      flex-direction: row-reverse;
    }
  }

  &--fullwidth {
    width: 100%;
  }

  &:not(:last-child) {
    margin-right: spacing();
  }

  &:not(:first-child) {
    margin: spacing('loose') 0;
  }

  &__optional {
    @include font-size(6);
    color: color('night', 'lightest');
    font-weight: $font-weight-light;
    text-transform: none;
  }
}
