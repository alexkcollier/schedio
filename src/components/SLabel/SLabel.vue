<template>
  <label :class="classList" class="label">
    <!-- Wrapped to keep optional on the right when reversed -->
    <span>
      {{ label }}
      <span v-if="isOptional" class="label__optional">&mdash;&nbsp;Optional</span>
    </span>
    <slot/>
  </label>
</template>

<script>
export default {
  name: 'SLabel',

  props: {
    label: {
      type: String,
      required: true
    },

    isOptional: {
      type: Boolean,
      default: true
    },

    isInline: {
      type: Boolean,
      default: false
    },

    isFullwidth: {
      type: Boolean,
      default: false
    },

    isReversed: {
      type: Boolean,
      default: false
    }
  },

  computed: {
    classList() {
      return {
        'label--inline': this.isInline,
        'label--fullwidth': this.isFullwidth,
        'label--reversed': this.isReversed
      }
    }
  },

  mounted() {
    if ((this.isFullwidth || this.isReversed) && !this.isInline) {
      console.error('`isFullwidth` and `isReversed` require `isInline`')
    }
  }
}
</script>
