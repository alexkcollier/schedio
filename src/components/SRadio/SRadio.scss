.radio {
  @include check-or-radio {
    border-radius: border-radius('circle');
  }

  & ~ &__check::after {
    background-color: color('green');
    border-radius: border-radius('circle');
    content: '';
    height: 0.65em;
    left: 50%;
    opacity: 0;
    position: absolute;
    top: 50%;
    transform: translateX(-50%) translateY(-50%);
    transition: all duration() ease;
    width: 0.65em;
  }

  &:checked ~ &__check::after {
    opacity: 1;
  }
}
