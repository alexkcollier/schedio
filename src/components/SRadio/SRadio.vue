<template>
  <SLabel
    :for="id"
    :label="label"
    :is-reversed="isReversed"
    :is-required="required"
    :hide-optional="hideOptional"
    is-inline
  >
    <input
      :id="id"
      v-bind="$attrs"
      :name="name"
      :required="required"
      class="radio"
      type="radio"
      v-on="listeners"
    >
    <span
      :class="{ 'radio__check--invalid': isInvalid }"
      class="radio__check"
    />
  </SLabel>
</template>

<script>
import InputControl from '@/mixins/InputControl.mixin'

export default {
  name: 'SRadio',

  mixins: [InputControl],
}
</script>
