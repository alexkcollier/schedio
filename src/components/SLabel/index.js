import SLabel from './SLabel.vue'

export { SLabel }

export default { install: Vue => Vue.component(SLabel.name, SLabel) }
