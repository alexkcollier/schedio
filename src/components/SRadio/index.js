import SRadio from './SRadio.vue';

export { SRadio };

export default { install: Vue => Vue.component(SRadio.name, SRadio) };
