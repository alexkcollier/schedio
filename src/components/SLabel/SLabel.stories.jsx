import { SHeading } from '@/components/SHeading'
import { SLabel } from '@/components/SLabel'
import PropList from '@@/docs/components/PropList'
import { boolean, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

storiesOf('Components/Inputs.SLabel', module)
  .addParameters({ jest: 'SLabel' })
  .add('Label', () => {
    return {
      props: {
        props: {
          default: () => ({
            label: text('Label text', 'Label text', 'Required Props'),
            isFullwidth: boolean('isFullwidth', false, 'Optional Props'),
            isInline: boolean('isInline', false, 'Optional Props'),
            isReversed: boolean('isReversed', false, 'Optional Props'),
            isRequired: boolean('iRrequired', false, 'Optional Props'),
            hideOptional: boolean('hide-optional', false, 'Optional Props')
          })
        },

        defaultSlot: {
          default: text('Default', 'SomeInputComponent', 'Slots')
        }
      },
      render(h) {
        const { props, defaultSlot } = this.$props
        return (
          <div>
            <SHeading level="1">Label</SHeading>
            <p>
              <code>SLabel</code> provides labels for all form inputs. It is included automatically
              for all input components in this library, and is designed to wrap components.
            </p>
            <p>
              By default, if a label is for an optional field, it will append '- Optional' to the
              label. This text can be hidden with the <code>hide-optional</code> prop.
            </p>

            <SHeading level="2">Example</SHeading>
            <SLabel {...{ props }}>
              <div style="text-transform: none; font-weight: 300;">
                <code>{'<' + defaultSlot + ' />'}</code>
              </div>
            </SLabel>

            <PropList component={SLabel} />
          </div>
        )
      }
    }
  })
