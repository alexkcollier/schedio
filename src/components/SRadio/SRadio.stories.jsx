import { SHeading } from '@/components/SHeading'
import { SRadio } from '@/components/SRadio'
import { withAttrsAsProps, withUnboundAttrs } from '@/mixins/stories/form-fields'
import PropList from '@/docs/PropList'
import StoryContainer from '@/docs/StoryContainer'
import StoryLink from '@/docs/StoryLink'
import { boolean, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

storiesOf('Components/SRadio', module)
  .addDecorator(withKnobs)
  .add('Radio', () => {
    const name = 'radio-group'

    return {
      props: {
        radios: {
          default: Array.from({ length: 3 }, (v, i) => ({
            props: {
              ...withAttrsAsProps({
                identifier: `radio-${i + 1}`,
                id: `radio-${i + 1}`,
                name,
                label: `Base Radio ${i + 1}`
              }),
              isReversed: boolean(`radio-${i + 1} isReversed`, false, 'Optional Props')
            },
            attrs: {
              ...withUnboundAttrs({ identifier: `radio-${i + 1}`, value: `radio-${i + 1}` }),
              'aria-labelledby': 'radio-example'
            }
          }))
        }
      },
      render(h) {
        return (
          <StoryContainer>
            <SHeading level="1">Radios</SHeading>
            <p>
              Radios provide the ability to make a single choice from a list of options. If the list
              is very long, consider using a <StoryLink to="Components/SSelect" />.
            </p>

            <SHeading level="2" id="radio-example">
              Example
            </SHeading>
            {this.radios.map(({ props, attrs }) => (
              <SRadio {...{ props }} {...{ attrs }} />
            ))}

            <PropList component={SRadio} />
          </StoryContainer>
        )
      }
    }
  })
