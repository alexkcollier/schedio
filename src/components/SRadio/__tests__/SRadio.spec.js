import { mount, shallowMount } from '@vue/test-utils'
import SRadio from '@/components/SRadio/SRadio.vue'
import SLabel from '@/components/SLabel/SLabel.vue'
import { axe, toHaveNoViolations } from 'jest-axe'

expect.extend(toHaveNoViolations)

describe('SRadio.vue', () => {
  const inputEvent = jest.fn()
  let wrapper
  let radio

  beforeEach(() => {
    wrapper = mount(SRadio, {
      propsData: {
        id: 'radio',
        name: 'radio',
        label: 'Radio',
      },
      listeners: {
        input: inputEvent,
      },
    })

    radio = wrapper.find('#radio')
  })

  it('renders correctly', async () => {
    expect(wrapper.html()).toMatchSnapshot()
    expect(await axe(wrapper.html())).toHaveNoViolations()
  })

  it('requires required props', () => {
    // spies on `console.error` without calling `console.error`
    const spy = jest.spyOn(global.console, 'error').mockImplementation(() => {})
    // mounting component without required props
    shallowMount(SRadio)
    expect(spy).toHaveBeenCalled()
    expect(spy.mock.calls[0][0]).toContain('[Vue warn]: Missing required prop')
  })

  it('can be checked', async () => {
    wrapper.trigger('click')
    expect(radio.element.checked).toBe(true)
    // check that radio emitted input/its value
    expect(wrapper.emitted('input')[0][0]).toBe(radio.element.value)
    expect(await axe(wrapper.html())).toHaveNoViolations()
  })

  it('can be a required input', () => {
    wrapper.setProps({ required: true })
    expect(radio.element.required).toBe(true)
  })

  it('can be invalid', () => {
    wrapper.setProps({ isInvalid: true })
    expect(radio.element.validity.valid).toBe(false)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be disabled', () => {
    const disabledWrapper = shallowMount(SRadio, {
      propsData: {
        id: 'radio',
        name: 'radio',
        label: 'Radio',
      },
      attrs: {
        disabled: true,
      },
    })

    // verifies disabled attr is passed through to radio input element
    expect(disabledWrapper.attributes('disabled')).toBeUndefined()
    expect(disabledWrapper.find('#radio').element.disabled).toBeTruthy()
  })

  it('can be reversed', () => {
    wrapper.setProps({ isReversed: true })
    expect(wrapper.find(SLabel).props('isReversed')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be grouped', () => {
    const groupWrapper = mount({
      components: {
        SRadio,
      },
      template: `
        <div>
          <SRadio
            name="radio-group"
            id="radio-1"
            label="radio-1"
            value="radio-1"
          />
          <SRadio
            name="radio-group"
            id="radio-2"
            label="radio-2"
            value="radio-2"
          />
        </div>
      `,
    })

    const radios = groupWrapper.findAll(SRadio)

    expect(radios.length).toBe(2)
    radios.trigger('click')
    expect(radios.at(0).emitted('input')[0][0]).toEqual('radio-1')
    expect(radios.at(1).emitted('input')[0][0]).toEqual('radio-2')
    expect(groupWrapper.html()).toMatchSnapshot()
  })

  it('can hide the optional tag', () => {
    wrapper.setProps({ hideOptional: true, required: false })
    expect(wrapper.text()).not.toContain('Optional')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
