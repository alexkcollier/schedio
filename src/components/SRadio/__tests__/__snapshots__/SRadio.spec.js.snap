// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SRadio.vue can be grouped 1`] = `
<div><label class="label label--inline" for="radio-1"><span>
    radio-1
    <!----></span> <input id="radio-1" name="radio-group" required="required" type="radio" class="radio" value="radio-1"> <span class="radio__check"></span></label> <label class="label label--inline" for="radio-2"><span>
    radio-2
    <!----></span> <input id="radio-2" name="radio-group" required="required" type="radio" class="radio" value="radio-2"> <span class="radio__check"></span></label></div>
`;

exports[`SRadio.vue can be invalid 1`] = `
<label class="label label--inline" for="radio"><span>
    Radio
    <!----></span> <input id="radio" name="radio" required="required" type="radio" class="radio"> <span class="radio__check radio__check--invalid"></span></label>
`;

exports[`SRadio.vue can be reversed 1`] = `
<label class="label label--inline label--reversed" for="radio"><span>
    Radio
    <!----></span> <input id="radio" name="radio" required="required" type="radio" class="radio"> <span class="radio__check"></span></label>
`;

exports[`SRadio.vue can hide the optional tag 1`] = `
<label class="label label--inline" for="radio"><span>
    Radio
    <!----></span> <input id="radio" name="radio" required="required" type="radio" class="radio"> <span class="radio__check"></span></label>
`;

exports[`SRadio.vue renders correctly 1`] = `
<label class="label label--inline" for="radio"><span>
    Radio
    <!----></span> <input id="radio" name="radio" required="required" type="radio" class="radio"> <span class="radio__check"></span></label>
`;
